src/solvers/smt2/smt2_conv.cpp

```diff
@@ -2672,7 +2672,14 @@ void smt2_convt::convert_struct(const struct_exprt &expr)
 
         // may need to flatten array-theory arrays in there
         if(op.type().id() == ID_array)
-          flatten_array(op);
+        {
+          const array_typet &array_type = to_array_type(op.type());
+          const auto &size_expr = array_type.size();
+          CHECK_RETURN(size_expr.id() == ID_constant);
+
+          if(numeric_cast_v<mp_integer>(to_constant_expr(size_expr)) != 0)
+            flatten_array(op);
+        }
```

Comment:
In principle, this check would now need to be done in front of every call to flatten array, to make that a precondition of that function.
I think this is the wrong approach. I would let flatten_array handle this case.